src/runtime-rs/crates/hypervisor/src/qemu/mod.rs

```diff
@@ -0,0 +1,129 @@
+mod inner;
+
+use crate::device::Device;
+use crate::hypervisor_persist::HypervisorState;
+use crate::Hypervisor;
+use crate::{HypervisorConfig, VcpuThreadIds};
+use inner::QemuInner;
+use kata_types::capabilities::Capabilities;
+
+use anyhow::Result;
+use async_trait::async_trait;
+
+use std::sync::Arc;
+use tokio::sync::RwLock;
+
+pub struct Qemu {
+    inner: Arc<RwLock<QemuInner>>,
+}
+
+impl Default for Qemu {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Qemu {
+    pub fn new() -> Self {
+        Self {
+            inner: Arc::new(RwLock::new(QemuInner::new())),
+        }
+    }
+
+    pub async fn set_hypervisor_config(&mut self, config: HypervisorConfig) {
+        let mut inner = self.inner.write().await;
+        inner.set_hypervisor_config(config)
+    }
+}
+
+#[async_trait]
+impl Hypervisor for Qemu {
+    async fn prepare_vm(&self, id: &str, netns: Option<String>) -> Result<()> {
```

Comment:
There is a lot of repeated boilerplate code to forward a request to `inner`. I wonder if this would not be worth defining a macro.

Reply:
Makes sense, that could be an interesting exercise. For now I did like the dragonball driver does.

Reply:
> There is a lot of repeated boilerplate code to forward a request to `inner`. I wonder if this would not be worth defining a macro.

TBH, I once notice this and tries to work on this. But currently, the macro expansion order is not determined in Rust, therefore, we cannot achieve that if we still want to use the `async_trait` proc macro.

I roughly have some ideas on this and I will try to "macroize" this part of the code soon.